aws_lambda_powertools/event_handler/api_gateway.py

```diff
@@ -0,0 +1,223 @@
+    ):
+        return self.route(rule, "PATCH", cors, compress, cache_control)
+
+    def route(
+        self,
+        rule: str,
+        method: str,
+        cors: Union[bool, CORSConfig] = False,
+        compress: bool = False,
+        cache_control: str = None,
+    ):
+        def register_resolver(func: Callable):
+            route = Route(method, self._compile_regex(rule), func, cors, compress, cache_control)
+            self._routes.append(route)
+            if route.cors:
+                if self._cors is None:
+                    self._cors = route.cors
+                self._cors_methods.add(route.method)
+            return func
+
+        return register_resolver
+
+    def resolve(self, event, context) -> Dict[str, Any]:
+        self.current_event = self._to_data_class(event)
+        self.lambda_context = context
+        route, args = self._find_route(self.current_event.http_method.upper(), self.current_event.path)
+        response = self.to_response(route.func(**args))
+
+        if route.cors:
+            response.add_cors(route.cors)
+        if route.cache_control:
+            response.add_cache_control(route.cache_control)
+        if route.compress and "gzip" in (self.current_event.get_header_value("accept-encoding") or ""):
+            response.compress()
+
+        return response.to_dict()
+
+    @staticmethod
+    def to_response(result: Union[Tuple[int, str, Union[bytes, str]], Dict, Response]) -> Response:
+        if isinstance(result, Response):
+            return result
+        elif isinstance(result, dict):
+            return Response(
+                status_code=200,
+                content_type="application/json",
+                body=json.dumps(result, separators=(",", ":"), cls=Encoder),
+            )
+        else:  # Tuple[int, str, Union[bytes, str]]
+            return Response(*result)
+
+    @staticmethod
+    def _compile_regex(rule: str):
+        rule_regex: str = re.sub(r"(<\w+>)", r"(?P\1.+)", rule)
+        return re.compile("^{}$".format(rule_regex))
+
+    def _to_data_class(self, event: Dict) -> BaseProxyEvent:
+        if self._proxy_type == ProxyEventType.http_api_v1:
+            return APIGatewayProxyEvent(event)
+        if self._proxy_type == ProxyEventType.http_api_v2:
+            return APIGatewayProxyEventV2(event)
+        return ALBEvent(event)
+
+    @staticmethod
+    def _preflight(allowed_methods: Set):
+        allowed_methods.add("OPTIONS")
+        headers = {"Access-Control-Allow-Methods": ",".join(sorted(allowed_methods))}
+        return Response(204, None, None, headers)
+
+    def _find_route(self, method: str, path: str) -> Tuple[Route, Dict]:
+        for route in self._routes:
+            if method != route.method:
+                continue
+            match: Optional[re.Match] = route.rule.match(path)
+            if match:
+                return route, match.groupdict()
+
+        if method == "OPTIONS" and self._cors is not None:
```

Comment:
@heitorlessa i noticed that Chalice does not do this for the api gateway, i am not exactly sure why not.

Reply:
:sigh: that's right - I totally forgot we do this at the API Gateway resource level, so it must be configured by the customer managing the API Gateway while the app code level is only responsible to ensure the response has the right headers.

You can safely remove that, sorry for the extra work

Reply:
@heitorlessa Not really. According to documentation: ["For a Lambda proxy integration or HTTP proxy integration, you can still set up the required OPTIONS response headers in API Gateway. However, your backend is responsible for returning the Access-Control-Allow-Origin and Access-Control-Allow-Headers headers, because a proxy integration doesn't return an integration response."](https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html) - Even testing with Chalice in local mode you get the right headers when using HTTP OPTIONS, so these are being created in the function.

Reply:
haha. should i stay or should i go. 

Either way, we would need docs explaining it :)